blockchain/src/state/read.rs

```diff
@@ -158,23 +165,26 @@ pub fn proof_from_trie_and_resolver(
     let storage_proof = if storage_slots.is_empty() {
         Vec::new()
     } else {
-        let mut storage = evm_db.storage_for(&address).ok()?;
+        let mut storage = storage_trie
+            .for_account_with_root(&address, &account_info.inner.storage_root)
+            .ok()?;
 
         storage_slots
             .iter()
-            .filter_map(|&index| {
+            .filter_map(|index| {
                 let key = keccak256::<[u8; 32]>(index.to_be_bytes());
-                storage.trie.get_proof(key.as_slice()).ok().map(|proof| {
-                    let value = storage.get(index);
+                storage.proof(key.as_slice()).ok().map(|proof| {
+                    let value = storage.get(index)?.unwrap_or_default();
 
-                    StorageProof {
-                        key: index.into(),
+                    Ok::<StorageProof, state::Error>(StorageProof {
+                        key: (*index).into(),
                         value,
                         proof: proof.into_iter().map(Into::into).collect(),
-                    }
+                    })
                 })
             })
-            .collect()
+            .collect::<Result<_, _>>()
+            .unwrap()
```

Comment:
Maybe we shouldn't `.unwrap()`

Reply:
Definitely.

At this point this function does not return result and a lot of the errors are ignored by returning `None`. But it's proving to cost some confusion when debugging as the function reports as "not found." while in some cases it really means "error."

So without error handling I'd say it's better to unwrap. While it does cause a panic, it makes it very clear where the error happened.

I think implementing proper result should be a part of a broader issue focused on error handling.

Reply:
I agree on the error handling needing to be separate, as there are a lot of things to revamp generally and I prefer PRs being semantic. Bulent's comment would also apply to e.g. panics [here](https://github.com/MovedNetwork/op-move/pull/306/files#diff-e1f6436696c104b1bc70e72f50808a691759740596e20125b88d777e1eb8f4feR580), though.